src/imitation/scripts/train_bc.py

```diff
@@ -0,0 +1,145 @@
+import logging
+import os.path as osp
+import pathlib
+from typing import Optional, Sequence, Union
+
+import gym
+from sacred.observers import FileStorageObserver
+from stable_baselines3.common import vec_env
+from torch.utils import data as th_data
+
+from imitation.algorithms import bc
+from imitation.data import rollout, types
+from imitation.scripts.config.train_bc import train_bc_ex
+from imitation.util import logger
+from imitation.util import sacred as sacred_util
+
+
+@train_bc_ex.main
+def train_bc(
+    _run,
+    expert_data_src: Union[types.AnyPath, Sequence[types.Trajectory]],
+    expert_data_src_format: str,
+    observation_space: gym.Space,
+    action_space: gym.Space,
+    batch_size: int,
+    # TODO(shwang): Doesn't currently accept Iterable[Mapping] or
+    #  types.TransitionsMinimal, unlike BC.__init__ or BC.set_expert_data_loader().
+    n_epochs: Optional[int],
+    n_batches: Optional[int],
+    n_expert_demos: int,
+    l2_weight: float,
+    optimizer_kwargs: dict,
+    log_dir: types.AnyPath,
+    venv: Optional[vec_env.VecEnv],
+    log_interval: int,
+    log_rollouts_n_episodes: int,
+    n_episodes_eval: int,
+) -> dict:
+    """Sacred interface to Behavioral Cloning.
+
+    Args:
+        expert_data_src: Either a path to pickled `Sequence[Trajectory]` or
+            `Sequence[Trajectory]`.
+        expert_data_src_format: Either "path" if `expert_data_src` is a path, or
+            "trajectory" if `expert_data_src` if `Sequence[Trajectory]`.
+        observation_space: The observation space corresponding to the expert data.
+        action_space: The action space corresponding to the expert data.
+        batch_size: Number of observation-action samples used in each BC update.
+        n_epochs: The total number of training epochs. Set exactly one of n_epochs and
+            n_batches.
+        n_batches: The total number of training batches. Set exactly one of n_epochs and
+            n_batches.
+        log_dir: Log output directory. Final policy is also saved in this directory as
+            "{log_dir}/final.pkl"
+        venv: If not None, then this VecEnv is used to generate rollout episodes for
+            evaluating policy performance during and after training.
+        log_interval: The number of updates in between logging various training
+            statistics to stdout and Tensorboard.
+        log_rollouts_n_episodes: The number of rollout episodes generated for
+            training statistics every `log_interval` updates. If `venv` is None or
+            this argument is nonpositive, then no rollouts are generated.
+        n_episodes_eval: The number of final evaluation rollout episodes, if `venv` is
+            provided. These rollouts are used to generate final statistics saved into
+            Sacred results, which can be compiled into a table by
+            `imitation.scripts.analyze.analyze_imitation`.
+    """
+    assert action_space is not None
```

Comment:
`assert` is meant for internal consistency checks, not input validation; note they will be disabled when running with `-O`. raise ValueError instead if missing.

Reply:
Makes sense, thanks